pkg/controller/knativeserving/openshift/openshift.go

```diff
@@ -581,3 +599,137 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
+		return nil
+	}
+
+	log.Info("exposing metric for installed knative version")
+	knativeVersion.WithLabelValues(version.Version, namespace).Set(float64(time.Now().Unix()))
+
+	// Get service knative-serving-operator in order to create service monitor
+	svcDetail, err := getOperatorService(namespace)
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			log.Info(fmt.Sprintf("service knative-serving-operator does not exist in %s namespace", namespace))
+			return nil
+		}
+		return err
+	}
+
+	// Add label openshift.io/cluster-monitoring to namespace
+	if err = addMonitoringLabelToNamespace(namespace); err != nil {
+		return err
+	}
+
+	// CreateServiceMonitors creates ServiceMonitors objects based on Service objects.
+	if _, err := metric.CreateServiceMonitors(cfg, namespace, []*v1.Service{svcDetail}); err != nil {
+		if strings.Contains(err.Error(), "already exists") {
+			log.Info(fmt.Sprintf("ServiceMonitor already exist for %s service", svcDetail.Name))
+			return nil
+		}
+		log.Error(err, "failed to create ServiceMonitor so removing added label from namespace")
+		if err = removeMonitoringLabelFromNamespace(namespace); err != nil {
+			return err
+		}
+		return err
+	}
+	log.Info("successfully created ServiceMonitor")
+	if err = createRole(namespace); err != nil {
+		return err
+	}
+	if err = createRoleBinding(namespace); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getOperatorService(ns string) (*corev1.Service, error) {
+	svc := &corev1.Service{}
+	err := api.Get(context.TODO(), types.NamespacedName{Name: "knative-serving-operator", Namespace: ns}, svc)
+	return svc, err
+}
+
+func getOperatorNamespace() (string, error) {
+	ns, found := os.LookupEnv(knativeServingInstalledNamespace)
+	if !found {
+		return "", fmt.Errorf("%s must be set", knativeServingInstalledNamespace)
+	}
+	return ns, nil
+}
+
+func addMonitoringLabelToNamespace(namespace string) error {
+	ns := &corev1.Namespace{}
+	if err := api.Get(context.TODO(), client.ObjectKey{Name: namespace}, ns); err != nil {
+		return err
+	}
+	if len(ns.Labels) == 0 {
+		ns.Labels = map[string]string{}
+	}
+	ns.Labels[monitoringLabel] = "true"
+	if err := api.Update(context.TODO(), ns); err != nil {
+		log.Error(err, fmt.Sprintf("could not add label %q to namespace %q", monitoringLabel, namespace))
+		return err
+	}
+	return nil
+}
+
+func removeMonitoringLabelFromNamespace(namespace string) error {
+	ns := &corev1.Namespace{}
+	if err := api.Get(context.TODO(), client.ObjectKey{Name: namespace}, ns); err != nil {
+		return err
+	}
+	if len(ns.Labels) == 0 {
+		return nil
+	}
+	if _, ok := ns.Labels[monitoringLabel]; ok {
+		delete(ns.Labels, monitoringLabel)
+	}
+	if err := api.Update(context.TODO(), ns); err != nil {
+		log.Error(err, fmt.Sprintf("could not remove label %q from namespace %q", monitoringLabel, namespace))
+		return err
+	}
+	return nil
+}
+
+func createRole(ns string) error {
+	role := &rbac.Role{}
```

Comment:
Why not adding all of the fields in here directly instead of mutating it in the following lines?